// cust-form-builder/esm5/lib/extValidService/external-validation.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
var ExternalValidationService = /** @class */ (function () {
    function ExternalValidationService() {
    }
    /**
     * @param {?} extValObj
     * @return {?}
     */
    ExternalValidationService.prototype.externalValidation = /**
     * @param {?} extValObj
     * @return {?}
     */
    function (extValObj) {
        //TODO: Not embedded action
        return JSON.stringify({
            target: extValObj.targetFields,
            source: extValObj.validationAction,
            label: extValObj.label
        });
    };
    ExternalValidationService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    ExternalValidationService.ctorParameters = function () { return []; };
    /** @nocollapse */ ExternalValidationService.ngInjectableDef = i0.defineInjectable({ factory: function ExternalValidationService_Factory() { return new ExternalValidationService(); }, token: ExternalValidationService, providedIn: "root" });
    return ExternalValidationService;
}());
export { ExternalValidationService };
