// cust-form-builder/esm2015/lib/extValidService/external-validation.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
export class ExternalValidationService {
    constructor() { }
    /**
     * @param {?} extValObj
     * @return {?}
     */
    externalValidation(extValObj) {
        //TODO: Not embedded action
        return JSON.stringify({
            target: extValObj.targetFields,
            source: extValObj.validationAction,
            label: extValObj.label
        });
    }
}
ExternalValidationService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
ExternalValidationService.ctorParameters = () => [];
/** @nocollapse */ ExternalValidationService.ngInjectableDef = i0.defineInjectable({ factory: function ExternalValidationService_Factory() { return new ExternalValidationService(); }, token: ExternalValidationService, providedIn: "root" });
