// cust-form-builder/esm5/lib/actionService/action.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
var ActionService = /** @class */ (function () {
    function ActionService() {
    }
    /**
     * @param {?} metadata
     * @param {?} originalMethod
     * @return {?}
     */
    ActionService.prototype.externalAction = /**
     * @param {?} metadata
     * @param {?} originalMethod
     * @return {?}
     */
    function (metadata, originalMethod) {
        var _a;
        /** @type {?} */
        var action = "/**\n";
        /** @type {?} */
        var parameters = [];
        /** @type {?} */
        var i = 0;
        for (var key in metadata.parameters) {
            action += "* @param {" + metadata.parameters[key] + "} " + key + "\n";
            parameters.push((_a = {}, _a[key] = metadata.args[i], _a));
            i++;
        }
        action += "* @returns {" + metadata.returnType + "}\n*/\n(" + originalMethod.toString() + ")();";
        return JSON.stringify({
            actionName: metadata.actionName,
            action: action,
            parameters: parameters,
            args: metadata.args
        });
    };
    /**
     * @param {?} metadata
     * @return {?}
     */
    ActionService.prototype.actionFromPackage = /**
     * @param {?} metadata
     * @return {?}
     */
    function (metadata) {
        return JSON.stringify({
            package: metadata.package,
            parameters: metadata.parameters
        });
    };
    ActionService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    ActionService.ctorParameters = function () { return []; };
    /** @nocollapse */ ActionService.ngInjectableDef = i0.defineInjectable({ factory: function ActionService_Factory() { return new ActionService(); }, token: ActionService, providedIn: "root" });
    return ActionService;
}());
export { ActionService };
