// cust-form-builder/esm5/lib/computedValueService/computed-value.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
var ComputedValueService = /** @class */ (function () {
    function ComputedValueService() {
    }
    /**
     * @param {...?} values
     * @return {?}
     */
    ComputedValueService.prototype.concat = /**
     * @param {...?} values
     * @return {?}
     */
    function () {
        var values = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            values[_i] = arguments[_i];
        }
        /** @type {?} */
        var result = {
            bind: {
                values: [],
                operator: "concat"
            }
        };
        for (var i in values) {
            result.bind.values.push(values[i]);
        }
        return JSON.stringify(result);
    };
    /**
     * @param {...?} values
     * @return {?}
     */
    ComputedValueService.prototype.add = /**
     * @param {...?} values
     * @return {?}
     */
    function () {
        var values = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            values[_i] = arguments[_i];
        }
        /** @type {?} */
        var obj = {
            bind: {
                values: [],
                operator: "add"
            }
        };
        for (var i in values) {
            obj.bind.values.push(values[i]);
        }
        return JSON.stringify(obj);
    };
    /**
     * @param {...?} values
     * @return {?}
     */
    ComputedValueService.prototype.substract = /**
     * @param {...?} values
     * @return {?}
     */
    function () {
        var values = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            values[_i] = arguments[_i];
        }
        /** @type {?} */
        var obj = {
            bind: {
                values: [],
                operator: "substract"
            }
        };
        for (var i in values) {
            obj.bind.values.push(values[i]);
        }
        return JSON.stringify(obj);
    };
    /**
     * @param {...?} values
     * @return {?}
     */
    ComputedValueService.prototype.multiply = /**
     * @param {...?} values
     * @return {?}
     */
    function () {
        var values = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            values[_i] = arguments[_i];
        }
        /** @type {?} */
        var obj = {
            bind: {
                values: [],
                operator: "multiply"
            }
        };
        for (var i in values) {
            obj.bind.values.push(values[i]);
        }
        return JSON.stringify(obj);
    };
    ComputedValueService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    ComputedValueService.ctorParameters = function () { return []; };
    /** @nocollapse */ ComputedValueService.ngInjectableDef = i0.defineInjectable({ factory: function ComputedValueService_Factory() { return new ComputedValueService(); }, token: ComputedValueService, providedIn: "root" });
    return ComputedValueService;
}());
export { ComputedValueService };
