// cust-form-builder/esm2015/lib/computedValueService/computed-value.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
export class ComputedValueService {
    constructor() { }
    /**
     * @param {...?} values
     * @return {?}
     */
    concat(...values) {
        /** @type {?} */
        var result = {
            bind: {
                values: [],
                operator: "concat"
            }
        };
        for (var i in values) {
            result.bind.values.push(values[i]);
        }
        return JSON.stringify(result);
    }
    /**
     * @param {...?} values
     * @return {?}
     */
    add(...values) {
        /** @type {?} */
        var obj = {
            bind: {
                values: [],
                operator: "add"
            }
        };
        for (var i in values) {
            obj.bind.values.push(values[i]);
        }
        return JSON.stringify(obj);
    }
    /**
     * @param {...?} values
     * @return {?}
     */
    substract(...values) {
        /** @type {?} */
        var obj = {
            bind: {
                values: [],
                operator: "substract"
            }
        };
        for (var i in values) {
            obj.bind.values.push(values[i]);
        }
        return JSON.stringify(obj);
    }
    /**
     * @param {...?} values
     * @return {?}
     */
    multiply(...values) {
        /** @type {?} */
        var obj = {
            bind: {
                values: [],
                operator: "multiply"
            }
        };
        for (var i in values) {
            obj.bind.values.push(values[i]);
        }
        return JSON.stringify(obj);
    }
}
ComputedValueService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
ComputedValueService.ctorParameters = () => [];
/** @nocollapse */ ComputedValueService.ngInjectableDef = i0.defineInjectable({ factory: function ComputedValueService_Factory() { return new ComputedValueService(); }, token: ComputedValueService, providedIn: "root" });
