// cust-form-builder/esm2015/lib/actionService/action.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
export class ActionService {
    constructor() { }
    /**
     * @param {?} metadata
     * @param {?} originalMethod
     * @return {?}
     */
    externalAction(metadata, originalMethod) {
        /** @type {?} */
        var action = "/**\n";
        /** @type {?} */
        var parameters = [];
        /** @type {?} */
        var i = 0;
        for (var key in metadata.parameters) {
            action += "* @param {" + metadata.parameters[key] + "} " + key + "\n";
            parameters.push({ [key]: metadata.args[i] });
            i++;
        }
        action += "* @returns {" + metadata.returnType + "}\n*/\n(" + originalMethod.toString() + ")();";
        return JSON.stringify({
            actionName: metadata.actionName,
            action: action,
            parameters: parameters,
            args: metadata.args
        });
    }
    /**
     * @param {?} metadata
     * @return {?}
     */
    actionFromPackage(metadata) {
        return JSON.stringify({
            package: metadata.package,
            parameters: metadata.parameters
        });
    }
}
ActionService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
ActionService.ctorParameters = () => [];
/** @nocollapse */ ActionService.ngInjectableDef = i0.defineInjectable({ factory: function ActionService_Factory() { return new ActionService(); }, token: ActionService, providedIn: "root" });
