// cust-form-builder/esm5/lib/regexService/regex.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
var RegexService = /** @class */ (function () {
    function RegexService() {
    }
    /**
     * @param {?} pattern
     * @param {?=} message
     * @return {?}
     */
    RegexService.prototype.regex = /**
     * @param {?} pattern
     * @param {?=} message
     * @return {?}
     */
    function (pattern, message) {
        return JSON.stringify({
            pattern: {
                value: pattern,
                message: message || ""
            }
        });
    };
    RegexService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    RegexService.ctorParameters = function () { return []; };
    /** @nocollapse */ RegexService.ngInjectableDef = i0.defineInjectable({ factory: function RegexService_Factory() { return new RegexService(); }, token: RegexService, providedIn: "root" });
    return RegexService;
}());
export { RegexService };
